man/man1/lpjml.1
.TH lpjml 1  "USER COMMANDS"
.SH NAME
lpjml \- Dynamic global vegetation model with managed land and river routing
.SH SYNOPSIS
.B lpjml
[\-h] [\-l] [\-v] [\-repo] [\-hash] [-vv] [-pedantic] [-print_noread] [-ofiles] [-param] [-nopp] [-pp cmd] [-fpe] [-image host[:inport[,outport]]] [-couple host[:port]] [\-wait time] [\-outpath \fIdir\fP]
[\-inpath \fIdir\fP] [\-restartpath \fIdir\fP]
[[\-Dmacro[=value]] [\-I\fIdir\fP] ...] \fIfilename\fP

.SH DESCRIPTION
The model simulates vegetation dynamics, hydrology and soil
organic matter dynamics on an area-averaged grid cell basis using
one-year time step. The name derives from the three locations \fBL\fPund-\fBP\fPotsdam-\fBJ\fPena but is no longer to be interpreted that way. Input parameters are monthly mean air
temperature, total precipitation, short and long wave radiation,
annual atmospheric CO2 concentration and soil texture class. The
simulation for each grid cell begins from "bare ground",
requiring a "spin up" (under non-transient climate) of c. 5000
years to develop equilibrium vegetation and soil structure. A restart
file can be created to save compute time for the spin-up phase. If the Message Passing Interface (MPI) library is available a parallel version of the program is built. If -DIMAGE flag is set in \fIMakefile.inc\fP coupling the IMAGE model via the TDT library is enabled.
.SH OPTIONS
.TP
\-h,\--help
Print a short help text.
.TP
\-l,\--license
Print the license file.
.TP
\-v,\--version
Print information about program version, operating system, compiler used, and compiler flags set.
.TP
\-repo
Print GIT repository.
.TP
\-hash
Print GIT hash.
.TP
\-vv
Verbosely print the actual values during reading of the configuration files.
.TP
\-pedantic
Stop on warnings.
.TP
\-print_noread
Print variable names not read from restart file.
.TP
\-ofiles
List only all available output variables.
.TP
\-param
Print LPJmL parameter for soils and PFTs and list of all available output files.
.TP
\-nopp
Disable preprocessing of configuration file by cpp or command specified by -pp option.
.TP
\-pp cmd
Set preprocessor program to cmd. Default is \fBcpp\fP.
.TP
\-fpe
Enable floating point exceptions. A core file is generated in case of an arithmetic error. Option is only available if \fBlpjml\fP has been compiled with -DWITH_FPE flag.
.TP
\-image host[:inport[,outport]]
Set host where IMAGE model is running. Default is localhost. inport and outport are the port numbers for ingoing and outgoing data. Default port numbers are 2225 and 2224, respectively. This option is only available for the IMAGE version of the code.
.TP
\-couple host[:port]
Set host where the coupled model is running. Default is localhost. Port is the port number for ingoing and outgoing data. Default port number is 2224.
.TP
\-wait time
Set time to wait for connection to coupled/IMAGE model. Time is measured in seconds.
.TP
\-outpath \fIdir\fP
Set the output directory path. The path is added to the output filenames if they do not contain an absolute path.
.TP
\-inpath \fIdir\fP
Set the input directory path. The path is added to the input filenames if they do not contain an absolute path.
.TP
\-restartpath \fIdir\fP
Set the restart directory path. The path is added to the restart filenames if they do not contain an absolute path.
.TP
\-Dmacro[=value]
Define macro for the preprocessor of the configuration file.
.TP
\-I\fIdir\fP
Define include directory for the preprocessor of the configuration file.
.TP
.I filename
Name of configuration file.
.SH EXAMPLES
.TP
Perform simulation with input from the /p/projects/lpjml file system:
.B lpjml
\-inpath /p/projects/lpjml lpjml.cjson
.PP
.TP
Start simulation from restart file with floating point exceptions enabled:
.B lpjml
\-DFROM_RESTART \-fpe lpjml.cjson
.PP
For running the MPI version of the program on more than one task the
.B lpjsubmit
script has to be used.
.SH ENVIRONMENT
.TP
LPJNOPP
Disables preprocessor if set to "true", same as -nopp option
.TP
LPJPEDANTIC
Enables pedantic mode if set to "true", same as -pedantic option
.TP
LPJPREP
Defines preprocessor command for LPJmL configuration file, default is \fBcpp\fP. Same as '-pp' option.
.TP
LPJROOT
Defines the root directory for LPJmL. This directory is added to the
include directories of the preprocessor. Invoking \fBlpj_paths.sh\fP or \fBlpj_paths.csh\fP will set this
environment variable correctly.
.TP
LPJCOUPLEDHOST
Sets the host where the coupled model is running. Same as '-couple' option.
.TP
LPJWAIT
Sets time to wait for connection to coupled/IMAGE model. Same as '-wait' option.
.TP
LPJIMAGE
Sets the host where IMAGE model is running. Same as '-image' option.
.TP
LPJINPATH
Path appended to the input filenames. Only done for filenames without absolute path. Same as '-inpath \fIdir\fP' option.
.TP
LPJOPTIONS
Runtime options for the preprocessor of LPJmL configuration files
.TP
LPJOUTPATH
Path appended to the output filenames. Only done for filenames without absolute path. Same as '-outpath \fIdir\fP' option.
.TP
LPJRESTARTPATH
Path appended to the restart filenames. Only done for filenames without absolute path. Same as '-restartpath \fIdir\fP' option.

.SH EXIT STATUS
.B lpjml
returns a zero exit status if the simulation finishes successfully.
Non zero is returned in case of failure.
.SH DIAGNOSTICS
On error the program prints the following message on stderr:

ERRORxxx: message, program terminated unsuccessfully

where xxx is the return code of the program. The following error codes are defined:
.TP
ERROR001
Error reading configuration. Check syntax of configuration file. External error
.TP
ERROR002
Error initializing input. Check whether all input files are present. External error
.TP
ERROR003
Error initializing grid, check for missing input files. External error
.TP
ERROR004
Invalid carbon balance. Internal error
.TP
ERROR005
Invalid water balance. Internal error
.TP
ERROR006
Negative discharge. Internal error
.TP
ERROR007
Negative fire probability. Internal error.
.TP
ERROR008
Negative soil moisture. Internal error
.TP
ERROR009
Error allocating memory. Rerun parallel program on more MPI tasks to reduce memory per task. External error
.TP
ERROR010
Negative stand fraction. Internal error.
.TP
ERROR011
Stand fraction sum error. Can be caused by invalid restart file. External error
.TP
ERROR012
List is empty in \fBdellistitem()\fP. Internal error.
.TP
ERROR013
Index out of range in \fBdellistitem()\fP. Internal error
.TP
ERROR014
Error in \fBnewlanduse()\fP. Can be caused by invalid restart file. External error
.TP
ERROR015
Invalid year in \fBgetco2()\fP. CO2 data file is too short. External error
.TP
ERROR016
Crop fraction >1. Internal error.
.TP
ERROR017
No natural stand for \fBdeforest()\fP. Internal error.
.TP
ERROR018
Wrong cultivation type. Internal error.
.TP
ERROR019
Floating point error occurred. Floating point exceptions will only be thrown if -fpe option is set. This is in particular useful if NaNs appear in the output files.  Internal error.
.TP
ERROR021
PFT list is not empty in \fBsetaside()\fP. Internal error.
.TP
ERROR022
Negative establishment rate. Internal error.
.TP
ERROR023
Output channel is broken. This error is only raised if LPJml is coupled to coupled model or IMAGE. It is usually caused by a premature end of the corresponding coupled model/IMAGE program. External error.
.TP
ERROR024
Error sending data to the IMAGE model. This error can only be raised if LPJmL is compiled with the -DIMAGE flag set in \fIMakefile.inc\fP. External error.
.TP
ERROR025
Error opening connection to IMAGE model. This error can only be raised if LPJmL is compiled with the -DIMAGE flag set in \fIMakefile.inc\fP. External error.
.TP
ERROR026
Not enough setaside stand created to put the reservoir. Internal error.
.TP
ERROR027
Forest left after deforestation. Internal error.
.TP
ERROR028
Outflow reservoir error. Internal error.
.TP
ERROR029
Error in permafrost module. Internal error.
.TP
ERROR030
Error in global water balance. Internal error.
.TP
ERROR031
Error in store climate function.
.TP
ERROR032
No FMS coupler supported.
.TP
ERROR033
Error initializing soil temperature
.TP
ERROR034
Invalid radiation model. Internal error
.TP
ERROR035
Negative soil NO3. Internal error
.TP
ERROR036
Negative soil NH4. Internal error
.TP
ERROR037
Invalid nitrogen balance. Internal error
.TP
ERROR038
Invalid climate data. External error
.TP
ERROR039
Invalid FPC value data. Internal error
.TP
ERROR040
Boundary check error. Internal error
.TP
ERROR041
Invalid time step in soiltemp(). Internal error
.TP
ERROR042
Cannot read external flow. External error
.TP
ERROR043
Opening connection to coupled model failed. External error
.TP
ERROR044
Invalid uniform temp sign value. Internal error
.TP
ERROR045
Error initializing output data. External error
.TP
ERROR046
Invalid crop PHU option. Internal error
.TP
ERROR047
Invalid fire danger index. Internal error
.TP
ERROR048
Gamma function error. Internal error
.TP
ERROR049
Wetland error. Internal error
.TP
ERROR050
Invalid hag_beta error. External error
.TP
ERROR051

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security


## [Unreleased]


## [6.1.0] - 2026-04-15

### Contributors

- author: Luke Oberhagemann (lukeober@pik-potsdam.de), Markus Drüke (Markus.Drueke@dwd.de), Maik Billing (billing@pik-potsdam.de), Werner von Bloh (bloh@pik-potsdam.de)
- code review: Stephen Wirth (wirth@pik-potsdam.de)

### Added

- Option `-inpath` added to `configure.sh` to set path for input files.
- Function `fprintclimbuf()` added in `climbuf.c` to print variables and is called in `fprintcell.c`.
- Function `fprintbuffer()` added in `buffer.c` to print variables and is called in `fprintclimbuf.c`.
- Lakes setting stored in restart file and setting compared to actual setting in `openrestart.c`.
- Utility `lpjcat` checks for identical lakes setting in restart files to concatenate.
- Urban stand added.
- Boolean `"relative_humidity"` added to lpjml configuration to use relative instead of absolute humidity for input.
- Fire related outputs and settings added:
```java
  "gsilivefuel" : true, /* livefuel calculated from growing season index (true/false) */
  "human_ign_prob" : false,   /* read human ignition probability from file */
  "max_firesize" : false,     /* read maximum fire size from file */
  "fireduration" : [{"stand" : "grassland", "duration" : [1,120] ,"ndayfire" : 1},
                    {"stand" : "natural", "duration" : [120,480], "ndayfire" : 3}], /* fire duration intervals and maximum days of fire for each stand */

  "prescribe_ignition" : false, /* read ignitions from file */
```
- Outputs `"pft_height"`, `"pft_agttop_litterc"`, `"littermoist"`, `"pft_phen"`, `"fireduration"`, `"firedurationdays"`, `"livegrass"`, `"dlm_livegrass"`, `"gsi_cum"`,
  `"gsi_diff"`, `"ndayfire"`, `"fuel"`, `"max_firesize"`, `"stand_fireduration"`, `"stand_burntarea"`, `"stand_fdi"`, `"stand_surface_fi`", `"fwi"`,
  `"human_ignition"`, `"lightning"`, `"surface_fi"`, `"ros"`, `"firesize"`, `"firedays"` added.

### Changed

- SPITFIRE fire model updated, multi-day fire implemented ([Oberhagemann et al. 2025](https://doi.org/10.5194/gmd-18-2021-2025)).
- The default fire model in `lpjml_config.cjson` is now set to `"spitfire_tmax"`.
- Keyword `"fire"` renamed to `"globfirm"`, `"no_fire"` to `"no"`.
- Simulation can be run without river routing from restart file with river routing. In this case a warning is printed.


## [6.0.7] - 2026-04-17

### Contributors

- author: Christoph Müller (cmueller@pik-potsdam.de)
- code review: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)

### Added

- `testcase_2cell` folder with standard inputs for a 2-cell riverbasin that can be used for testing the model and the `globalflux.csv` and `globalflux_spinup.csv` outputs for comparison
- `make test` option that automatically runs the 2-cell testcase
- maintaining the `globalflux.csv` and `globalflux_spinup.csv` files as author task for merge requests

### Changed

- modified `.gitignore` to allow for `*.clm` and `*.dat` input files in the `testcase_2cell` folder
- modified `Makefile` for improved parallel compilation of `all` target


## [6.0.6] - 2026-03-25

### Contributors

author: Stephen Wirth (wirth@pik-potsdam.de), Werner von Bloh (bloh@pik-potsdam.de)
code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)

### Added

- DEBUG print statements for negative input to infiltration in `daily_agriculture()`, `daily_agriculture_grass()`, `daily_agriculture_tree()`, `daily_biomass_grass()`, `daily_biomass_tree()`, `daily_woodplantation()`, and `daily_natural()`
- Missing regridding of climate data added to the `regridlpj` utility

### Changed

- Replace hardcoded `Sphagnum_moss` index (13) with `strcmp(config->pftpar[p].name,"Sphagnum moss")` in 5 landuse files

lpjml.js
/**************************************************************************************/
/**                                                                                \n**/
/**                   l  p  j  m  l  .  j  s                                       \n**/
/**                                                                                \n**/
/** Default configuration file for LPJmL C Version 5.7.10                           \n**/
/**                                                                                \n**/
/** Configuration file is divided into five sections:                              \n**/
/**                                                                                \n**/
/**  I.   Simulation description and type section                                  \n**/
/**  II.  Input parameter section                                                  \n**/
/**  III. Input data section                                                       \n**/
/**  IV.  Output data section                                                      \n**/
/**  V.   Run settings section                                                     \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

//#define BMGR_BROWN    /* enables brown harvest of biomass grass, instead of green harvest (default) */

{   /* LPJmL configuration in JSON format */

/*===================================================================*/
/*  I. Simulation description and type section                       */
/*===================================================================*/

  "sim_name" : "LPJmL Run", /* Simulation description */
  "sim_id"   : "lpjml",     /* LPJML Simulation type with managed land use */
  "coupled_model" : null,   /* Coupled model: null (no model coupled), string (name of coupled model) */
  "start_coupling": null,   /* Start year of model coupling: null (start_coupling is set to firstyear if coupled_model != null), int (start year of coupling) */
  "version"  : "5.7",       /* LPJmL version expected */
  "random_prec" : true,     /* Random weather generator for precipitation enabled */
  "random_seed" : 2,        /* seed for random number generator */
  "radiation" : "radiation",/* other options: "cloudiness", "radiation", "radiation_swonly", "radiation_lwdown" */
  "fire" : "fire",          /* fire disturbance enabled, other options: NO_FIRE, FIRE, SPITFIRE, SPITFIRE_TMAX (for GLDAS input data) */
  "fire_on_grassland" : false, /* enable fire on grassland for Spitfire */
  "fdi" : "nesterov",       /* different fire danger index formulations: "wvpd" (needs GLDAS input data), "nesterov" */
  "firewood" : false,
  "gsi_phenology" : true,   /* GSI phenology enabled */
  "transp_suction_fcn" : false, /* enable transpiration suction function (true/false) */
  "with_lakes" : true,      /* enable lakes (true/false) */
  "river_routing" : true,
  "extflow" : false,
  "permafrost" : true,
  "johansen" : true,
  "soilpar_option" : "no_fixed_soilpar", /* other options "no_fixed_soilpar", "fixed_soilpar", "prescribed_soilpar" */
  "with_nitrogen" : "lim", /* other options: "no", "lim", "unlim" */
  "store_climate" : true, /* store climate data in spin-up phase */
  "landfrac_from_file" : true, /* read cell area from file (true/false) */
  "shuffle_spinup_climate" : true, /* shuffle spinup climate */
  "fix_climate" : false,                /* fix climate after specified year */
  "fix_climate_year" : 1901,            /* year after climate is fixed */
  "fix_climate_interval" : [1901,1930],
  "fix_climate_shuffle" : true,          /* randomly shuffle climate in the interval */
  "fix_deposition_with_climate" : false, /* fix N deposition same as climate  */
  "fix_deposition" : false,              /* fix N deposition after specified year */
  "fix_deposition_year" : 1901,          /* year after deposition is fixed */
  "fix_deposition_interval" : [1901,1930],
  "fix_deposition_shuffle" : true,       /* randomly shuffle depositions in the interval */
  "fix_landuse" : false,                 /* fix land use after specfied year */
  "fix_landuse_year" : 1901,             /* year after land use is fixed */
  "fix_co2" : false,                     /* fix atmospheric CO2  after specfied year */
  "fix_co2_year" : 1901,                 /* year after CO2 is fixed */
#ifdef FROM_RESTART
  "new_seed" : false, /* read random seed from restart file */
  "population" : false,
  "landuse" : "yes", /* other options: "no", "yes", "const", "all_crops", "only_crops" */
  "landuse_year_const" : 2000, /* set landuse year for "const" case */
  "reservoir" : true,
  "wateruse" : "yes",  /* other options: "no", "yes", "all" */
  "equilsoil" : false,
#else
  "equilsoil" : true,
  "population" : false,
  "landuse" : "no",
  "reservoir" : false,
  "wateruse" : "no",
#endif
  "prescribe_burntarea" : false,
  "prescribe_landcover" : "no_landcover", /* NO_LANDCOVER, LANDCOVERFPC, LANDCOVEREST */
  "sowing_date_option" : "fixed_sdate",   /* NO_FIXED_SDATE, FIXED_SDATE, PRESCRIBED_SDATE */
  "sdate_fixyear" : 1970,               /* year in which sowing dates shall be fixed */
  "intercrop" : false,                  /* intercrops on setaside */
  "residue_treatment" : "fixed_residue_remove", /* residue options: READ_RESIDUE_DATA, NO_RESIDUE_REMOVE, FIXED_RESIDUE_REMOVE (uses param residues_in_soil) */ 
  "residues_fire" : false,              /* fire in residuals */
  "irrigation" : "lim",                 /* other options: "no", "lim", "pot", "all" */
  "laimax_interpolate" : "laimax_par",  /* laimax values from manage parameter file, */
                                        /* other options: LAIMAX_CFT, CONST_LAI_MAX, LAIMAX_INTERPOLATE */
  "tillage_type" : "all",               /* Options: "all" (all agr. cells tilled), "no" (no cells tilled) and "read" (tillage dataset used) */
  "till_startyear" : 1850,              /* year in which tillage should start */
  "black_fallow" : false,               /* simulation with black fallow on PNV */
  "pft_residue" : "temperate cereals",
  "no_ndeposition" : false,             /* turn off atmospheric N deposition */
  "rw_manage" : false,                  /* rain water management */
  "laimax" : 5,                         /* maximum LAI for CONST_LAI_MAX */
  "fertilizer_input" : "yes",           /* enable fertilizer input, other options: "no", "yes", "auto" */
  "manure_input" : true,                /* enable manure input */
  "fix_fertilization" : false,          /* fix fertilizer input */
  "others_to_crop" : true,              /* move PFT type others into PFT crop, cft_tropic for tropical,  cft_temp for temperate */
  "grazing" : "default",                /* default grazing type, other options : "default", "mowing", "ext", "int", "livestock", "none" */
  "grazing_others" : "default",         /* default grazing type for others, other options : "default", "mowing", "ext", "int", "livestock", "none" */
  "cft_temp" : "temperate cereals",
  "cft_tropic" : "maize",
  "grassonly" : false,                  /* set all cropland including others to zero but keep managed grasslands */
  "luc_timber" : true,                  /* land-use change timber */
  "grassland_fixed_pft" : false,
  "grass_harvest_options" : false,
  "prescribe_lsuha" : false,
  "mowing_days" : [152, 335],          /* Mowing days for grassland if grass harvest options are ser */
  "crop_resp_fix" : false,             /* variable C:N ratio for crop respiration */
                                       /* for MAgPIE runs, turn off dynamic C:N ratio dependent respiration,
                                          which reduces yields at high N inputs */
  "crop_phu_option" : "new",
  "cropsheatfrost" : false,
  "double_harvest" : true,
  "npp_controlled_bnf" : true,

/*===================================================================*/
/*  II. Input parameter section                                      */
/*===================================================================*/

#include "param.js"    /* Input parameter file */

/*===================================================================*/
/*  III. Input data section                                          */
/*===================================================================*/

#include "input_crumonthly.js"    /* Input files of CRU dataset */

/*===================================================================*/
/*  IV. Output data section                                          */
/*===================================================================*/

#ifdef WITH_GRIDBASED
  "grid_scaled" : true, /* PFT-specific outputs scaled by stand->frac */
#define SUFFIX grid.bin
#else
  "grid_scaled" : false,
#define SUFFIX pft.bin
#endif

  "output_metafile" : true, /* no json metafile created */
  "float_grid" : false,      /* set datatype of grid file to float (TRUE/FALSE) */

#define mkstr(s) xstr(s) /* putting string in quotation marks */
#define xstr(s) #s

#ifdef FROM_RESTART

  "output" :
  [

/*
ID                               Fmt                        filename
-------------------------------- ------------------------- ----------------------------- */
    { "id" : "grid",             "file" : { "fmt" : "raw", "name" : "output/grid.bin" }},
    { "id" : "terr_area",        "file" : { "fmt" : "raw", "name" : "output/terr_area.bin" }},
    { "id" : "land_area",        "file" : { "fmt" : "raw", "name" : "output/land_area.bin" }},
    { "id" : "lake_area",        "file" : { "fmt" : "raw", "name" : "output/lake_area.bin" }},
    { "id" : "fpc",              "file" : { "fmt" : "raw", "name" : "output/fpc.bin" }},
    { "id" : "globalflux",       "file" : { "fmt" : "txt", "name" : "output/globalflux.csv"}},
    { "id" : "npp",              "file" : { "fmt" : "raw", "name" : "output/mnpp.bin"}},
    { "id" : "gpp",              "file" : { "fmt" : "raw", "name" : "output/mgpp.bin"}},
    { "id" : "rh",               "file" : { "fmt" : "raw", "name" : "output/mrh.bin"}},
    { "id" : "rd",               "file" : { "fmt" : "raw", "name" : "output/rd.bin"}},
    { "id" : "transp",           "file" : { "fmt" : "raw", "name" : "output/mtransp.bin"}},
    { "id" : "runoff",           "file" : { "fmt" : "raw", "name" : "output/mrunoff.bin"}},
    { "id" : "evap",             "file" : { "fmt" : "raw", "name" : "output/mevap.bin"}},
    { "id" : "interc",           "file" : { "fmt" : "raw", "name" : "output/minterc.bin"}},
    { "id" : "swc1",             "file" : { "fmt" : "raw", "name" : "output/mswc1.bin"}},
    { "id" : "swc2",             "file" : { "fmt" : "raw", "name" : "output/mswc2.bin"}},
    { "id" : "vegc",             "file" : { "fmt" : "raw", "name" : "output/vegc.bin"}},
    { "id" : "soilc",            "file" : { "fmt" : "raw", "name" : "output/soilc.bin"}},
    { "id" : "litc",             "file" : { "fmt" : "raw", "name" : "output/litc.bin"}},
    { "id" : "flux_estabc",      "file" : { "fmt" : "raw", "name" : "output/flux_estab.bin"}},
    { "id" : "vegn",             "file" : { "fmt" : "raw", "name" : "output/vegn.bin"}},
    { "id" : "soiln",            "file" : { "fmt" : "raw", "name" : "output/soiln.bin"}},
    { "id" : "litn",             "file" : { "fmt" : "raw", "name" : "output/litn.bin"}},
    { "id" : "leaching",         "file" : { "fmt" : "raw", "name" : "output/mleaching.bin"}},
    { "id" : "n2o_denit",        "file" : { "fmt" : "raw", "name" : "output/mn2o_denit.bin"}},
    { "id" : "n2o_nit",          "file" : { "fmt" : "raw", "name" : "output/mn2o_nit.bin"}},
    { "id" : "n2_emis",          "file" : { "fmt" : "raw", "name" : "output/mn2_emis.bin"}},
    { "id" : "bnf",              "file" : { "fmt" : "raw", "name" : "output/mbnf.bin"}},
    { "id" : "n_immo",           "file" : { "fmt" : "raw", "name" : "output/mn_immo.bin"}},
    { "id" : "nfert_agr",        "file" : { "fmt" : "raw", "name" : "output/nfert_agr.bin"}},
    { "id" : "nfert_mg",     "file" : { "fmt" : "raw", "name" : "output/nfert_mg.bin"}},
    { "id" : "firen",            "file" : { "fmt" : "raw", "name" : "output/firen.bin"}},
    { "id" : "n_mineralization", "file" : { "fmt" : "raw", "name" : "output/mn_mineralization.bin"}},
    { "id" : "n_volatilization", "file" : { "fmt" : "raw", "name" : "output/mn_volatilization.bin"}},
    { "id" : "ndepos",           "file" : { "fmt" : "raw", "name" : "output/ndepos.bin"}},
    { "id" : "pft_laimax",       "file" : { "fmt" : "raw", "name" : "output/pft_laimax.bin"}},
#ifdef WITH_SPITFIRE
    { "id" : "firec",            "file" : { "fmt" : "raw", "timestep" : "monthly" , "unit" : "gC/m2/month", "name" : "output/mfirec.bin"}},
    { "id" : "nfire",            "file" : { "fmt" : "raw", "name" : "output/mnfire.bin"}},
    { "id" : "burntarea",        "file" : { "fmt" : "raw", "name" : "output/mburnt_area.bin"}},
#else
    { "id" : "firec",            "file" : { "fmt" : "raw", "name" : "output/firec.bin"}},
#endif
    { "id" : "discharge",        "file" : { "fmt" : "raw", "name" : "output/mdischarge.bin"}},
    { "id" : "wateramount",      "file" : { "fmt" : "raw", "name" : "output/mwateramount.bin"}},
    { "id" : "harvestc",         "file" : { "fmt" : "raw", "name" : "output/flux_harvest.bin"}},
    { "id" : "sdate",            "file" : { "fmt" : "raw", "name" : "output/sdate.bin"}},
    { "id" : "pft_harvestc",     "file" : { "fmt" : "raw", "name" : mkstr(output/pft_harvest.SUFFIX)}},
    { "id" : "cftfrac",          "file" : { "fmt" : "raw", "name" : "output/cftfrac.bin"}},
    { "id" : "pet",              "file" : { "fmt" : "raw", "name" : "output/mpet.bin"}},
    { "id" : "albedo",           "file" : { "fmt" : "raw", "name" : "output/malbedo.bin"}},
    { "id" : "maxthaw_depth",    "file" : { "fmt" : "raw", "name" : "output/maxthaw_depth.bin"}},
    { "id" : "perc",             "file" : { "fmt" : "raw", "name" : "output/mperc.bin"}},
    { "id" : "soiltemp1",        "file" : { "fmt" : "raw", "name" : "output/msoiltemp1.bin"}},
    { "id" : "return_flow_b",    "file" : { "fmt" : "raw", "name" : "output/mreturn_flow_b.bin"}},
    { "id" : "transp_b",         "file" : { "fmt" : "raw", "name" : "output/mtransp_b.bin"}},
    { "id" : "evap_b",           "file" : { "fmt" : "raw", "name" : "output/mevap_b.bin"}},
    { "id" : "interc_b",         "file" : { "fmt" : "raw", "name" : "output/mintec_b.bin"}},
    { "id" : "prod_turnover",    "file" : { "fmt" : "raw", "name" : "output/prod_turnover.bin"}},
    { "id" : "deforest_emis",    "file" : { "fmt" : "raw", "name" : "output/deforest_emis.bin"}},
    { "id" : "conv_loss_evap",   "file" : { "fmt" : "raw", "name" : "output/aconv_loss_evap.bin"}},
    { "id" : "conv_loss_drain",  "file" : { "fmt" : "raw", "name" : "output/aconv_loss_drain.bin"}},
    { "id" : "bnf_agr",          "file" : { "fmt" : "raw", "name" : "output/bnf_agr.bin"}},
    { "id" : "bnf_mg",           "file" : { "fmt" : "raw", "name" : "output/bnf_mg.bin"}},
    { "id" : "nfert_agr",        "file" : { "fmt" : "raw", "name" : "output/nfert_agr.bin"}},
    { "id" : "nmanure_agr",      "file" : { "fmt" : "raw", "name" : "output/nmanure_agr.bin"}},
    { "id" : "ndepo_agr",        "file" : { "fmt" : "raw", "name" : "output/ndepo_agr.bin"}},
    { "id" : "ndepo_mg",         "file" : { "fmt" : "raw", "name" : "output/ndepo_mg.bin"}},
    { "id" : "nmineralization_agr", "file" : { "fmt" : "raw", "name" : "output/nmineralization_agr.bin"}},
    { "id" : "nimmobilization_agr", "file" : { "fmt" : "raw", "name" : "output/nimmobilization_agr.bin"}},
    { "id" : "nuptake_agr",      "file" : { "fmt" : "raw", "name" : "output/nuptake_agr.bin"}},
    { "id" : "nleaching_agr",    "file" : { "fmt" : "raw", "name" : "output/nleaching_agr.bin"}},
    { "id" : "n2o_denit_agr",    "file" : { "fmt" : "raw", "name" : "output/n2o_denit_agr.bin"}},
    { "id" : "n2o_nit_agr",      "file" : { "fmt" : "raw", "name" : "output/n2o_nit_agr.bin"}},
    { "id" : "nh3_agr",          "file" : { "fmt" : "raw", "name" : "output/nh3_agr.bin"}},
    { "id" : "n2_agr",           "file" : { "fmt" : "raw", "name" : "output/n2_agr.bin"}},
    { "id" : "harvestn_agr",     "file" : { "fmt" : "raw", "name" : "output/harvestn_agr.bin"}},
    { "id" : "harvestc_agr",     "file" : { "fmt" : "raw", "name" : "output/harvestc_agr.bin"}},
    { "id" : "seedn_agr",        "file" : { "fmt" : "raw", "name" : "output/seedn_agr.bin"}},
    { "id" : "cellfrac_agr",     "file" : { "fmt" : "raw", "name" : "output/cellfrac_agr.bin"}},
    { "id" : "litfalln_agr",     "file" : { "fmt" : "raw", "name" : "output/litfalln_agr.bin"}},
    { "id" : "rootmoist",        "file" : { "fmt" : "raw", "name" : "output/rootmoist.bin"}},
    { "id" : "uptakec_mgrass",   "file" : { "fmt" : "raw", "name" : "output/uptakec_mgrass.bin"}},
    { "id" : "yieldn_mgrass",    "file" : { "fmt" : "raw", "name" : "output/yieldn_mgrass.bin"}},
    { "id" : "no3_leaching_mgrass", "file" : { "fmt" : "raw", "name" : "output/no3_leaching_mgrass.bin"}},
    { "id" : "n2o_nit_mgrass",   "file" : { "fmt" : "raw", "name" : "output/n2o_nit_mgrass.bin"}},
    { "id" : "n2o_denit_mgrass", "file" : { "fmt" : "raw", "name" : "output/n2o_denit_mgrass.bin"}},
    { "id" : "n2_mgrass",        "file" : { "fmt" : "raw", "name" : "output/n2_mgrass.bin"}},
    { "id" : "nh3_mgrass",       "file" : { "fmt" : "raw", "name" : "output/nh3_mgrass.bin"}},
    { "id" : "irrig",         "file" : { "fmt" : "raw", "name" : "output/irrig.bin"}},
    { "id" : "harvestn",         "file" : { "fmt" : "raw", "name" : "output/flux_harvestn.bin"}},
    { "id" : "flux_estabn_mg","file" : { "fmt" : "raw", "name" : "output/flux_estabn_mg.bin"}},
    { "id" : "pft_npp",          "file" : { "fmt" : "raw", "name" : "output/pft_npp.bin"}},
    { "id" : "timber_harvestc",          "file" : { "fmt" : "raw", "name" : "output/timber_harvestc.bin"}},
    { "id" : "prec",          "file" : { "fmt" : "raw", "name" : "output/prec.bin"}},
    { "id" : "temp",          "file" : { "fmt" : "raw", "name" : "output/temp.bin"}}
/*------------------------------ ------------------------- ------------------------------- */
  ],

#else

  "output" :
  [
    { "id" : "globalflux",       "file" : { "fmt" : "txt", "name" : "output/globalflux_spinup.csv"}}
  ],
#endif

/*===================================================================*/
/*  V. Run settings section                                          */
/*===================================================================*/

  "startgrid" : "all", /* 27410, 67208 60400 47284 47293 47277 all grid cells */
  "endgrid"   : "all",
#ifdef CHECKPOINT
  "checkpoint_filename" : "restart/restart_checkpoint.lpj", /* filename of checkpoint file */
#endif

#ifndef FROM_RESTART

  "nspinup" : 3500,  /* spinup years */
  "nspinyear" : 30,  /* cycle length during spinup (yr) */
  "firstyear": 1901, /* first year of simulation */
  "lastyear" : 1901, /* last year of simulation */
  "restart" :  false, /* start from restart file */
  "outputyear" : -1599,
  "write_restart" : true, /* create restart file: the last year of simulation=restart-year */
  "write_restart_filename" : "restart/restart_1840_nv_stdfire.lpj", /* filename of restart file */
  "restart_year": 1840 /* write restart at year */

#else

  "nspinup" : 390,   /* spinup years */
  "nspinyear" : 30,  /* cycle length during spinup (yr)*/
  "firstyear": 1901, /* first year of simulation */
  "lastyear" : 2011, /* last year of simulation */
  "outputyear": 1901, /* first year output is written  */
  "restart" :  true, /* start from restart file */
  "restart_filename" : "restart/restart_1840_nv_stdfire.lpj", /* filename of restart file */
  "write_restart" : true, /* create restart file */
  "write_restart_filename" : "restart/restart_1900_crop_stdfire.lpj", /* filename of restart file */
  "restart_year": 1900 /* write restart at year */

#endif
}
